cmd/tier.go

```diff
@@ -202,22 +207,75 @@ func (config *TierConfigMgr) GetDriver(tierName string) (d warmBackend, err erro
 	return d, nil
 }
 
-func saveGlobalTierConfig() error {
-	b, err := globalTierConfigMgr.Bytes()
+// configReader returns a PutObjReader and ObjectOptions needed to save config
+// using a PutObject API. PutObjReader encrypts json encoded tier configurations
+// if KMS is enabled, otherwise simply yields the json encoded bytes as is.
+// Similarly, ObjectOptions value depends on KMS' status.
+func (config *TierConfigMgr) configReader() (*PutObjReader, *ObjectOptions, error) {
+	b, err := config.Bytes()
 	if err != nil {
-		return err
+		return nil, nil, err
+	}
+
+	payloadSize := int64(len(b))
+	br := bytes.NewReader(b)
+	hr, err := hash.NewReader(br, payloadSize, "", "", payloadSize, false)
+	if err != nil {
+		return nil, nil, err
+	}
+	if GlobalKMS == nil {
+		return NewPutObjReader(hr, nil, nil), &ObjectOptions{}, nil
+	}
+
+	// Note: Local variables with names ek, oek, etc are named in line with
+	// acronyms defined here -
+	// https://github.com/minio/minio/blob/master/docs/security/README.md#acronyms
+
+	// Encrypt json encoded tier configurations
+	metadata := make(map[string]string)
+	kmsContext := crypto.Context{minioMetaBucket: minioMetaBucket, TierConfigPath: TierConfigPath}
+	ek, encEK, err := GlobalKMS.GenerateKey(GlobalKMS.DefaultKeyID(), kmsContext)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	oek := crypto.GenerateKey(ek, rand.Reader)
+	encOEK := oek.Seal(ek, crypto.GenerateIV(rand.Reader), crypto.S3.String(), minioMetaBucket, TierConfigPath)
+	crypto.S3.CreateMetadata(metadata, GlobalKMS.DefaultKeyID(), encEK, encOEK)
+	encBr, err := sio.EncryptReader(br, sio.Config{Key: oek[:], MinVersion: sio.Version20})
+	if err != nil {
+		return nil, nil, err
+	}
+
+	info := ObjectInfo{
+		Size: payloadSize,
+	}
+	encSize := info.EncryptedSize()
+
+	encHr, err := hash.NewReader(encBr, encSize, "", "", payloadSize, false)
+	if err != nil {
+		return nil, nil, err
 	}
-	r, err := hash.NewReader(bytes.NewReader(b), int64(len(b)), "", "", int64(len(b)), false)
+	opts := &ObjectOptions{
+		ServerSideEncryption: encrypt.NewSSE(),
+		UserDefined:          metadata,
+		MTime:                UTCNow(),
```

Comment:
MTime does not need to be set - it anyway defaults to UTCNow() if unset in erasure layer